Feature/boltdb

pawanrawal opened the pull request:
Wal and other components of posting (get and set), now use Bolt. On preliminary analysis, its slow for the 21M dataset.

manishrjain commented:
Review status: 0 of 14 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[posting/index.go, line 230 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/529#-KbHm5ZWpptD_Zh_UwAl:-KbHm5ZXUtQrNQ4Wy7zW:b-8ulu45) ([raw file](https://github.com/dgraph-io/dgraph/blob/587f4010f328313bda8220ae35145af86bbda812/posting/index.go#L230)):*
> ```Go
> 	batchCh := make(chan []byte, 1000)
> 	errCh := make(chan error, 1)
> 	go batchDelete(batchCh, errCh)
> ```

Why is this a goroutine?

---

*[posting/list.go, line 185 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/529#-KbHmIvYbJJvUhIMr5GB:-KbHmIvZvAoKiVQvyjU8:b-lqzs1q) ([raw file](https://github.com/dgraph-io/dgraph/blob/587f4010f328313bda8220ae35145af86bbda812/posting/list.go#L185)):*
> ```Go
> 		l.pstore.View(func(tx *bolt.Tx) error {
> 			b := tx.Bucket([]byte("data"))
> 			x.Checkf(plist.Unmarshal(b.Get(l.key)), "Unable to Unmarshal PostingList from store")
> ```

Error handling. Doesn't this get triggered?

---

*[raftwal/wal.go, line 104 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/529#-KbHnqIVIG1TQYX_4uz6:-KbHnqIW1FrTWw9Po-eh:b7b15ke) ([raw file](https://github.com/dgraph-io/dgraph/blob/587f4010f328313bda8220ae35145af86bbda812/raftwal/wal.go#L104)):*
> ```Go
> 
> 	var batch Batch
> 	err := w.wals.View(func(tx *bolt.Tx) error {
> ```

Why do this in 2 different steps -- read then write? Why not just do an update, and do whatever you need to do straight away?

Also, in general, acquiring a read lock then using that information in a separate block with write lock is a recipe for failure. The write block would always need to redetermine whatever conditions the read block determined.

---

*[worker/predicate.go, line 40 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/529#-KbHou9qVn-M7sXD50uO:-KbHou9rl0xrnKkZ4Kjw:b-tzzm61) ([raw file](https://github.com/dgraph-io/dgraph/blob/587f4010f328313bda8220ae35145af86bbda812/worker/predicate.go#L40)):*
> ```Go
> func writeBatch(ctx context.Context, kv chan *task.KV, che chan error) {
> 	for i := range kv {
> 		if err := pstore.Batch(func(tx *bolt.Tx) error {
> ```

This is sure to be slow. You want to acquire the lock once and push out as much as you can in that lock. Though, this isn't in critical path, so won't change the numbers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/529)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Review status: 0 of 14 files reviewed at latest revision, 4 unresolved discussions.

---

*[posting/index.go, line 230 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/529#-KbHm5ZWpptD_Zh_UwAl:-KbHv7lMtrC25d7dwZxa:bss4a2b) ([raw file](https://github.com/dgraph-io/dgraph/blob/587f4010f328313bda8220ae35145af86bbda812/posting/index.go#L230)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Why is this a goroutine?
</blockquote></details>

Removed.

---

*[posting/list.go, line 185 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/529#-KbHmIvYbJJvUhIMr5GB:-KbHsVHSgTw1Ej5h8OVW:b-7kdfoe) ([raw file](https://github.com/dgraph-io/dgraph/blob/587f4010f328313bda8220ae35145af86bbda812/posting/list.go#L185)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Error handling. Doesn't this get triggered?
</blockquote></details>

Doing it now, didn't get triggered earlier.

---

*[raftwal/wal.go, line 104 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/529#-KbHnqIVIG1TQYX_4uz6:-KbHsdeqtVY0Ox0lZi2C:b2br1l4) ([raw file](https://github.com/dgraph-io/dgraph/blob/587f4010f328313bda8220ae35145af86bbda812/raftwal/wal.go#L104)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Why do this in 2 different steps -- read then write? Why not just do an update, and do whatever you need to do straight away?

Also, in general, acquiring a read lock then using that information in a separate block with write lock is a recipe for failure. The write block would always need to redetermine whatever conditions the read block determined.
</blockquote></details>

Was doing it in a single step earlier and it was slow. You are right this doesn't make sense, reverted the changes.

---

*[worker/predicate.go, line 40 at r1](https://reviewable.io:443/reviews/dgraph-io/dgraph/529#-KbHou9qVn-M7sXD50uO:-KbHtHjV_7sKFXQcRu5L:byr4flj) ([raw file](https://github.com/dgraph-io/dgraph/blob/587f4010f328313bda8220ae35145af86bbda812/worker/predicate.go#L40)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

This is sure to be slow. You want to acquire the lock once and push out as much as you can in that lock. Though, this isn't in critical path, so won't change the numbers.
</blockquote></details>

Right.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/529)*
<!-- Sent from Reviewable.io -->
